tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -43,18 +47,25 @@ namespace disc_ral {
 namespace {
 
 template <typename LHLO_OpTy>
-LogicalResult elemwiseLowerHelper(
-    OpBuilder& b, Location loc, Operation* op, Value output_linear_index,
-    const ShapeConstraintAnalysis* shape_constraint_analysis, int vector_size) {
+LogicalResult elemwiseLowerHelper(OpBuilder& b, Location loc, Operation* op,
+                                  Value output_linear_index,
+                                  const ShapeAnalysis* shape_analysis,
+                                  int vector_size, LowerConfig* lower_config) {
   if (!isa<LHLO_OpTy>(op) || !op->hasTrait<mlir::OpTrait::Elementwise>())
     return failure();
 
   Value result_memref = cast<lmhlo::LmhloOp>(op).getResultBuffer();
   Value memref = result_memref;
-  if (shape_constraint_analysis) {
-    Value leader_memref =
-        shape_constraint_analysis->GetLeaderValueWithSameShape(result_memref);
-    if (leader_memref != nullptr) memref = leader_memref;
+  if (shape_analysis) {
+    auto parent_fusion = op->getParentOp();
+    while ((parent_fusion != nullptr) && !isa<mhlo::FusionOp>(parent_fusion)) {
+      parent_fusion = parent_fusion->getParentOp();
+    }
```

Comment:
1, the while can be replaced by getParentOfType<mhlo::FusionOp>()
2, if elemwiseLowerHelper() is called by a non-fusion node, then shape_analysis will be nullptr, so why do we need to check if it's inside a fusion?

Reply:
Thanks for the comment.
I added the nullptr-check to guarantee that this single function is robust, no matter how it is called (without the prior knowledge that a non-fusion lead to an empty shapeAnalysis). But I just realize we indeed do not need this check, because if fusion is nullptr, `GetLeaderValueWithSameShapeInFusion' will return nullptr.

Reply:
Updated.